### DISCUSSION TITLE
How to get locations and glyph indexes of ligatures in shaped output?

### DISCUSSION THREAD

**Discussion by smolck** [harfbuzz/harfbuzz#3264] [Q&A] [answered]
Basically, given let's say the string `-> --> => >=`, assuming each set of chars separated by spaces forms a ligature in the font of choice, is there a way to shape said text and then get the ranges in the text where the ligatures are, and also the glyph indexes for each of those? For example in this text, it'd give 0-1 and then a glyph index for the `->` ligature, 3-5 and then a glyph index for the `-->` ligature, etc. (and if the text was a combination of ligatures and non-ligatures, i.e. `hello ->`, it would only give the locations and glyph indexes of the ligatures, so just `->` in this case).

After shaping it's possible to get the `codepoint` and `cluster` values for each glyph, but without knowing what's a ligature and where those ligatures are, I don't think I can get very far with only that?

**Comment by khaledhosny** [comment, marked as answer]
https://harfbuzz.github.io/working-with-harfbuzz-clusters.html

**Reply by smolck** [reply]
Oh TIL, thank you! So if I'm understanding correctly, the way to do what I want is to loop over the glyph infos resulting from shaping a piece of text, and then wherever a series of glyphs have the same cluster value they belong to the same ligature?

**Reply by khaledhosny** [reply]
If multiple glyphs have the same cluster number, then they belong the the same input character i.e. the opposite of ligature.

For ligatures you will find gaps in the cluster numbers, e.g.: `0,1,5` means glyph 0 corresponds to character 0, glyph 1 to characters 1 to 4, glyph 2 to character 5.

`0,0,1` means glyphs 0 and 1 correspond to the character 0, glyph 2 to character 1.

Both can be combined, e.g.: `0,1,1,1,5` means glyphs 1 to 3 correspond to characters 1 to 4.

**Reply by smolck** [reply]
Looked over https://harfbuzz.github.io/a-clustering-example-for-levels-0-and-1.html and the concrete example there cleared things up; thanks for the help!
